webapp/src/app.tsx

```diff
@@ -1845,11 +1816,49 @@ export class ProjectView
         if (devIndicator) devIndicator.clear()
     }
 
+    shouldStartSimulator(): boolean {
+        const hasHome = !pxt.shell.isControllerMode();
+        if (!hasHome) return true;
+        return !this.state.home;
+    }
+
     isSimulatorRunning() {
         return this.state.running;
     }
 
-    runSimulator(opts: compiler.CompileOptions = {}, clickTrigger?: boolean) {
+    restartSimulator(debug?: boolean) {
+        this.stopSimulator();
+        this.startSimulator(debug);
+    }
+
+    startSimulator(debug?: boolean, clickTrigger?: boolean) {
+        pxt.tickEvent('simulator.start');
+        if (!this.shouldStartSimulator() || (this.runToken && this.runToken.isRunning())) {
+            pxt.log("Ignoring call to start simulator, either already running or we shouldn't start.");
+            return;
+        }
+        this.saveFileAsync()
+            .then(() => this.runSimulator({ debug, clickTrigger }))
+    }
+
+    stopSimulator(unload?: boolean, dontCancel?: boolean) {
+        pxt.tickEvent('simulator.stop')
+        if (!dontCancel && this.runToken) this.runToken.cancel()
```

Comment:
Because runSimulator calls stopSimulator when it begins, which would cancel the token midway through. We only want to cancel if a different action called stopSimulator.

Reply:
We could alternately call simulator.stop directly and we can probably avoid the tick event here when its an internal call to stop the simulator.

Reply:
The token is internal sauce, refactor stop into an internal method.